File: 0130-surrounded-regions/0130-surrounded-regions.cpp
// TC- O(m*n)
// SC- O(m*n)

class Solution {
private:
    int rows, cols;
    void dfs(vector<vector<char>>& board, int i, int j){
        // using # to mark all unsurrounded regions instead of keeping track of visited 
        board[i][j] = '#';
        vector<int> directions = {-1, 0, 1, 0, -1};
        for(int d=0; d<directions.size()-1; d++){
            int new_i = i+directions[d];
            int new_j = j+directions[d+1];
            if(new_i>=0 and new_j>=0 and new_i<rows and new_j<cols and board[new_i][new_j]=='O')
                dfs(board, new_i, new_j);
        }
        return;
    }
public:
    void solve(vector<vector<char>>& board) {
        rows = board.size();
        cols = board[0].size();

        // find unsurrounded regions
        for(int r=0; r<rows; r++){
            for(int c=0;c<cols; c++){
                if((r==0 or c==0 or r==rows-1 or c==cols-1) and board[r][c]=='O'){
                    dfs(board, r, c);
                }
            }
        }

        // Change # into O and O into X
        for(int r=0; r<rows; r++){
            for(int c=0;c<cols; c++){
                if(board[r][c]=='#')
                    board[r][c]= 'O';
                else if(board[r][c]=='O')
                    board[r][c]= 'X';
            }
        }
        return;
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.py
# TC- O(mn)
# SC- O(mn)

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows, cols = len(board), len(board[0])

        def dfs(r, c):
            board[r][c] = "I"
            dirs = [-1, 0, 1, 0, -1]

            for d in range(len(dirs)-1):
                next_r = r+dirs[d]
                next_c = c+dirs[d+1]
                if 0<=next_r<rows and 0<=next_c<cols and board[next_r][next_c]=="O":
                    dfs(next_r, next_c)

        for i in range(cols):
            if board[0][i]=="O":
                dfs(0, i)
        for i in range(cols):
            if board[rows-1][i]=="O":
                dfs(rows-1, i)
        for i in range(rows):
            if board[i][0]=="O":
                dfs(i, 0)
        for i in range(rows):
            if board[i][cols-1]=="O":
                dfs(i, cols-1)
        
        for i in range(rows):
            for j in range(cols):
                if board[i][j]=="O":
                    board[i][j] = "X"
                elif board[i][j]=="I":
                    board[i][j] = "O"
        